ledger/acctonline.go

```diff
@@ -644,81 +647,97 @@ func (ao *onlineAccounts) lookupOnlineAccountData(rnd basics.Round, addr basics.
-						err = fmt.Errorf("failed to write history of acct %s for round %d into online accounts cache", data.addr.String(), data.updRound)
-						return ledgercore.OnlineAccountData{}, err
-					}
+	// No updates of this account in the in-memory deltas; use on-disk DB.
+	// As an optimization, we avoid creating
+	// a separate transaction here, and directly use a prepared SQL query
+	// against the database.
+	var persistedData persistedOnlineAccountData
+	persistedData, err = ao.accountsq.lookupOnline(addr, rnd)
+	if err != nil || persistedData.rowid == 0 {
+		// no such online account, return empty
+		return ledgercore.OnlineAccountData{}, err
+	}
+	// Now we load the entire history of this account to fill the onlineAccountsCache, so that the
+	// next lookup for this online account will not hit the on-disk DB.
+	//
+	// lookupOnlineHistory fetches the account DB round from the acctrounds table (validThrough) to
+	// distinguish between different cases involving the last-observed value of ao.cachedDBRoundOnline.
+	// 1. Updates to ao.onlineAccountsCache happen with ao.accountsMu taken below, as well as in postCommit()
+	// 2. If we started reading the history (lookupOnlineHistory)
+	//   1. before commitRound or while it is running => OK, read what is in DB and then add new entries in postCommit
+	//     * if commitRound deletes some history after, the cache has additional entries and updRound comparison gets a right value
+	//   2. after commitRound but before postCommit => OK, read full history, ignore the update from postCommit in writeFront's updRound comparison
+	//   3. after postCommit => OK, postCommit does not add new entry with writeFrontIfExist, but here all the full history is loaded
+	var persistedDataHistory []persistedOnlineAccountData
+	var validThrough basics.Round
+	persistedDataHistory, validThrough, err = ao.accountsq.lookupOnlineHistory(addr)
+	if err != nil || len(persistedDataHistory) == 0 {
+		return ledgercore.OnlineAccountData{}, err
+	}
+	// 3. After we finished reading the history (lookupOnlineHistory), either
+	//   1. The DB round has not advanced (validThrough == currentDbRound) => OK
+	//   2. after commitRound but before postCommit (currentDbRound >= ao.cachedDBRoundOnline && currentDeltaLen == len(ao.deltas)) => OK
+	//      the cache gets populated and postCommit updates the new entry
+	//   3. after commitRound and after postCommit => problem
+	//      postCommit does not add a new entry, but the cache that would get constructed would miss the latest entry, retry
+	// In order to resolve this lookupOnlineHistory returns dbRound value (as validThrough) and determine what happened
+	// So handle cases 3.1 and 3.2 here, and 3.3 below
+	out, err = ao.lookupOnlineAccountDataHandleData(addr, currentDbRound, currentDeltaLen, &persistedData, persistedDataHistory, validThrough, rewardsProto, rewardsLevel)
+	if err == errNotDoneYet {
+		// case 3.3: retry (for loop iterates and queries again)
+		goto slowDbRetry
```

Comment:
why a goto vs a for loop? They are pretty similar ... is there a reason for one over the other?